In [ ]:
import pandas as pd
from sklearn.linear_model import LinearRegression,LogisticRegression
from sklearn.model_selection import train_test_split

In [ ]:
df=pd.read_csv("/content/loan_default_risk_dataset.csv")
df

Retirement_Age  Debt_Amount  Monthly_Savings  Loan_Default_Risk
0              60.0      2996.52          2378.49                  0
1              66.4      4137.23          1538.92                  1
2              58.5     19865.75          2434.80                  1
3              49.8     16855.70          2677.82                  1
4              67.3      7902.37          2206.72                  0
..              ...          ...              ...                ...
295            71.9      4393.93          2324.35                  0
296            56.0      5861.44          1916.44                  1
297            61.9      2503.60          2073.36                  0
298            56.9      4754.51          2603.25                  0
299            58.7     12743.30          1591.53                  1

[300 rows x 4 columns]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Retirement_Age     300 non-null    float64
 1   Debt_Amount        299 non-null    float64
 2   Monthly_Savings    298 non-null    float64
 3   Loan_Default_Risk  300 non-null    int64  
dtypes: float64(3), int64(1)
memory usage: 9.5 KB


In [ ]:
debtmean=df['Debt_Amount'].mean()
df['Debt_Amount']=df['Debt_Amount'].fillna(debtmean)

In [ ]:
monthmean=df['Monthly_Savings'].mean()
df['Monthly_Savings']=df['Monthly_Savings'].fillna(monthmean)

In [ ]:
x=df.drop('Loan_Default_Risk',axis=1)
y=df['Loan_Default_Risk']

In [ ]:
xtrain,xtest,ytrain,ytest=train_test_split(x,y,train_size=0.8,random_state=42)

In [ ]:
model=LogisticRegression()
model.fit(xtrain,ytrain)

LogisticRegression()

In [ ]:
trainscore=model.score(xtrain,ytrain)
trainscore

1.0

In [ ]:
testscore=model.score(xtest,ytest)
testscore

0.9833333333333333

In [ ]:
df.corr()['Loan_Default_Risk']

Retirement_Age      -0.421008
Debt_Amount          0.342097
Monthly_Savings     -0.497029
Loan_Default_Risk    1.000000
Name: Loan_Default_Risk, dtype: float64

**MULTI CLASS LABELS**

In [ ]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris

In [ ]:
data=load_iris(as_frame=True)
x=data.data
y=data.target

In [ ]:
xtrain,xtest,ytrain,ytest=train_test_split(x,y,train_size=0.8,random_state=42)

In [ ]:
model=LogisticRegression()
'''
important parameter:-
1)penalty:-values=l1-lasso,l2=ridge,elasticnet,None
2)tol:-=0.0001 if last(previous weights)-current weights=0.0001 stops the iteration
3)c:Float=1,(regularization strength)
4)class_weight:-None,balanced
5)random_state:-takes constant weights every times
6)solver:-['lbfgs','liblinear','newton-cg','sag','saga']='lbfgs'
'liblinear'->small dataset
'lbfgs'->small to medium dataset
'newton-cg'->small to medium->used when we need fast convergence(fastly it reduce the weights)
'saga'->large dataset Text based or sentence based scenario(TF-IDF vectorizer/NLP)
'sag'->large dataset for all the scenario
(optimzation technique which helps for gradient descent to reduce the weights)
7)max_iter:-Int=100(maximum number of iterations to reduce the weights in gradient descent)
8)multi_class:-['auto','ovr'(one vs rest all),'multinomial'(softmax)]='auto'(binary|sigmoid)
9)l1_ratio->use when penalty is elasticnet
if l1_ration is 0->100% l2
if l1_ratio is 1->100% l1
if l1_ratio is 0.3->30% l1 ,70% l2
multinomial=softmax function
auto->both-binary=sigmoid,multiclass=softmax
Gradient reduces weights using penalty and solver
'''

"\nimportant parameter:-\n1)penalty:-values=l1-lasso,l2=ridge,elasticnet,None\n2)tol:-=0.0001 if last(previous weights)-current weights=0.0001 stops the iteration\n3)c:Float=1,(regularization strength)\n4)class_weight:-None,balanced\n5)random_state:-takes constant weights every times\n6)solver:-['lbfgs','liblinear','newton-cg','sag','saga']='lbfgs'\n'liblinear'->small dataset\n'lbfgs'->small to medium dataset\n'newton-cg'->small to medium->used when we need fast convergence(fastly it reduce the weights)\n'saga'->large dataset Text based or sentence based scenario(TF-IDF vectorizer/NLP)\n'sag'->large dataset for all the scenario\n(optimzation technique which helps for gradient descent to reduce the weights)\n7)max_iter:-Int=100(maximum number of iterations to reduce the weights in gradient descent)\n8)multi_class:-['auto','ovr'(one vs rest all),'multinomial'(softmax)]='auto'(binary|sigmoid)\n9)l1_ratio->use when penalty is elasticnet\nmultinomial=softmax function\nauto->both-binary=sigm

In [ ]:
model=LogisticRegression()
model.fit(xtrain,ytrain)

LogisticRegression()

In [ ]:
model.score(xtrain,ytrain)

0.975

In [ ]:
model.score(xtest,ytest)

1.0

In [ ]:
ytrain.value_counts()

target
1    41
0    40
2    39
Name: count, dtype: int64

In [ ]:
data.target_names

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [ ]:
model=LogisticRegression(n_jobs=1)
model.fit(xtrain,ytrain)

LogisticRegression(n_jobs=1)

In [ ]:
ypred=model.predict(xtrain)

In [ ]:
from sklearn.metrics import confusion_matrix,precision_score,recall_score,accuracy_score,classification_report
confusion_matrix(ytrain,ypred)

array([[40,  0,  0],
       [ 0, 38,  3],
       [ 0,  0, 39]])

* setosa[0]:out of all 40 values it is correctly predicted as 40
* versicolor[1]:out of all 41 it is correctly predicted as 38 and wrongly predicted as 3
* virginica[2]:out of all 39 it is correctly predicted as 39


* when average !='binary' pos_label is not considered
* macro:is not working for imbalance class labels

In [ ]:
print(precision_score(ytrain,ypred,average='macro'))
print(precision_score(ytrain,ypred,average='micro'))
print(precision_score(ytrain,ypred,average='weighted'))
print(precision_score(ytrain,ypred,average=None))

0.9761904761904763
0.975
0.9767857142857144
[1.         1.         0.92857143]


* precision_micro:individual performance,class labels are balanced
   * p0:TP/TP+FP=40/40+0=1
   * p1:TP/TP+FP=38/38+0=1
   * p2:TP/TP+FP=39/39+3=0.92
   * summationkTP=40+38+39/summationk(TP+FP)=40+3+42=0.975
* precision_macro:overall weighted class labels performance,when class labels are imbalanced
  * p0:1
  * p1:1
  * p2:0.92
  * p0+p1+p2/3=1+1+0.92/3=0.9761
* precision_weighted:with respect to class labels how values are correctly predicted-conclusion based on class labels
  * supports->class label values
      * p0=40
      * p1=41
      * p2=39
  * summationkprecision x supportsk/total supports
     * 1x40+1x41+0.92x39/120=0.976
* precision=None:when we want individual precision score


In [ ]:
print(recall_score(ytrain,ypred,average='micro'))
print(recall_score(ytrain,ypred,average='macro'))
print(recall_score(ytrain,ypred,average='weighted'))

0.975
0.975609756097561
0.975


In [ ]:
print(accuracy_score(ytrain,ypred))

0.975


In [ ]:
print(classification_report(ytrain,ypred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        40
           1       1.00      0.93      0.96        41
           2       0.93      1.00      0.96        39

    accuracy                           0.97       120
   macro avg       0.98      0.98      0.97       120
weighted avg       0.98      0.97      0.97       120

